Change desktop setting as 11sp4 KDE upgrade to 15SP3 gnome

tinawang123 opened the pull request:
Need send ret key to handle login for KDE desktop

- ticket: https://progress.opensuse.org/issues/88467
- Verification run: https://openqa.suse.de/tests/5689699#step/first_boot/4

okurz reviewed (changes requested):
This PR was already in and reverted. To give reviewers context I suggest to reference the original PR, the revert, explain the problem, why it was reverted and what your approach is doing differently now.

Also, 
The commit message could have a bit more details, e.g. what issue is fixed, why this needs to change, to which versions of which product it applies, link to a bug or a fate entry, the choices you made, etc. Also see https://commit.style/ or http://chris.beams.io/posts/git-commit/ as a helpful guide how to write good commit messages. I am stating that often for many people. I do not know if you have read this suggestion by me already, likely you already have :) The suggestion is a general best practice so also valid for any other further changes, not just this one

tinawang123 commented:
Before it can impact opensuse upgrade cases, Now I added flavor settting check, Only 'migration' cases will use this condition.
It will not impact opensuse cases.

tinawang123 commented:
@okurz, Please help to check this PR

Vogtinator reviewed (changes requested):
I suggest to just change `DESKTOP` to `gnome` during the migration, as that's what happens on the SUT.

Otherwise you'll have to sprinkle this extra condition to all occurences of `check_var('DESKTOP', 'gnome')` at some point...

Vogtinator commented:
The image the migration is triggered from is a SLE 11 SP4 with KDE, so it has `DESKTOP` set to `KDE`. During the migration, GNOME is installed and replaces KDE. So afterwards it boots into GNOME, but openQA still has `DESKTOP` as `KDE` and thus runs `KDE` specific code.

So the failure is not KDE related at all, the issue is that openQA runs KDE code against the GNOME login prompt.

tinawang123 commented:
Verified: https://openqa.suse.de/tests/5741865#

Vogtinator commented on tests/installation/logs_from_installation_system.pm:
```diff
@@ -59,6 +59,7 @@ sub run {
     # We don't change network setup here, so should work
     # We don't parse logs unless it's detect_yast2_failures scenario
     $self->save_upload_y2logs(no_ntwrk_recovery => 1, skip_logs_investigation => !get_var('ASSERT_Y2LOGS'));
+    set_var('DESKTOP', 'gnome') if (get_var('KDE2GNOME'));
```
This belongs more into `migration/post_upgrade.pm` than this module, which is only for collecting logs.

Is a new variable necessary? Would `is_sle('<12') && check_var('DESKTOP', 'kde')` be enough?

Vogtinator reviewed (changes requested):
Commit message and PR title need adjustment

tinawang123 commented:
@Vogtinator Updated PR

Vogtinator reviewed (changes requested):
Code looks good, but commit and PR subject are still wrong

Vogtinator commented:
> As post_upgrade module is after first_boot. It cannot impact first_boot.
> So I changed the setting at first_boot module.

Hm, what about `tests/migration/version_switch_upgrade_target.pm`? That already has code to handle the switch from KDE, so that could just be extended.

tinawang123 commented:
Verified: https://openqa.suse.de/tests/5746192#

tinawang123 commented:
> > As post_upgrade module is after first_boot. It cannot impact first_boot.
> > So I changed the setting at first_boot module.
> 
> Hm, what about `tests/migration/version_switch_upgrade_target.pm`? That already has code to handle the switch from KDE, so that could just be extended.

It's a good place. I changed the setting at this file. It can work now. Thanks a lot

Vogtinator commented on tests/migration/version_switch_upgrade_target.pm:
```diff
@@ -50,6 +50,9 @@ sub run {
     set_var('DM_NEEDS_USERNAME', '1') if (check_var('DESKTOP', 'KDE') && is_sle('15+') && (get_var('ADDONURL', '') !~ /phub/));
 
+    # Reset DESKTOP after upgrade as desktop change
+    set_var('DESKTOP', 'gnome') if (is_sle('<12', get_var('HDDVERSION')) && check_var('DESKTOP', 'kde'));
```
There is already a very similar condition a line up, so that can be reused:

```
# In SLE 15, KDE was dropped, after migration the default desktop is GNOME
# KDE has moved to Package Hub, it will stay installed with SLE 15 if PackageHub is available
if (check_var('DESKTOP', 'KDE') && is_sle('15+') && (get_var('ADDONURL', '') !~ /phub/)) {
    set_var('DM_NEEDS_USERNAME', '1');
    set_var('DESKTOP', 'gnome');
}
```

I'm a bit confused by the two assignments to `DM_NEEDS_USERNAME` which can conflict with each other and also the comment which says `after migration the default desktop is XDM`, which doesn't seem to be true.

tinawang123 replied:
This is a little different with my condition. For my case, After upgrade, it may upgrade to sles12 or sles15... I only need check base system is 11sp4 with KDE. After upgrade, the desktop will change to gnome.  For "if (check_var('DESKTOP', 'KDE') && is_sle('15+') && (get_var('ADDONURL', '') !~ /phub/))" this condition. It means the desktop will stay to use 'KDE', it won't change to gnome

tinawang123 replied:
my understand: For desktop is KDE, it need set DM_NEEDS_USERNAME=1. But after upgrade, the desktop changed to gnome, it should set DM_NEEDS_USERNAME=0.

Vogtinator replied:
> For my case, After upgrade, it may upgrade to sles12 or sles15...

Are there still migration tests from 11SP4+KDE to SLE12? AFAICT that would also switch to gnome, so setting `DESKTOP` to `gnome` in that case would still be correct. Just changing `is_sle('15+')` to `is_sle('12+')` would be needed then.

tinawang123 replied:
I checked the case and code. Those code not used any more. So, I deleted those code and change the desktop setting to the first condition.

Vogtinator replied:
> I checked the case and code. Those code not used any more. So, I deleted those code and change the desktop setting to the first condition.

Are you sure? If you upgrade from 11 SP4 to 15 with packagehub enabled, you won't end up with GNOME.

tinawang123 replied:
I add the condition to check whether sles15 include packagehub module.

tinawang123 replied:
I add the condition to check whether sles15 include packagehub module.

tinawang123 commented:
Verified: https://openqa.suse.de/tests/5754189

Vogtinator commented on tests/migration/version_switch_upgrade_target.pm:
```diff
@@ -43,12 +43,11 @@ sub run {
-    set_var('DM_NEEDS_USERNAME', '1') if (check_var('DESKTOP', 'KDE') && is_sle('15+') && (get_var('ADDONURL', '') !~ /phub/));
+    # Reset DESKTOP after upgrade as desktop change
+    if (is_sle('=11-SP4', get_var('HDDVERSION')) && check_var('DM_NEEDS_USERNAME', '1')) {
+        set_var('DM_NEEDS_USERNAME', '0');
```
Previously, `DM_NEEDS_USERNAME` was set to `1` on migration (line 51), but now it's set to `0`. Is that intentional?
Might have to be coupled with the desktop change

tinawang123 replied:
Only 11sp4 should set DM_NEEDS_USERNAME=1. After upgrade, the system become sles11sp+, it should set DM_NEEDS_USERNAME=0. This is expected.

Vogtinator replied:
Previously it was set to `1` after upgrade though

tinawang123 replied:
That code don't work. I checked the case vars.json file, the setting DM_NEEDS_USERNAME=0. So it didn't set to 1 after upgrade.

tinawang123 replied:
Checked openqa code:
sub check_var {
    my ($var, $val) = @_;
    return 1 if (defined $bmwqemu::vars{$var} && $bmwqemu::vars{$var} eq $val);
    return 0;
}
It will compare check_vars('DESKTOP','KDE'). As it is upper, our setting is DESKTOP=kde,  So those code never be used. It don't impact others.

tinawang123 commented:
@Vogtinator  Do you have any comment about this PR?

Vogtinator reviewed (approved):
LGTM!
